### PR TITLE
fix build error with gcc 12+

### LINKED ISSUE #127920: Pytorch build from source failed with GCC 12.3
### 🐛 Describe the bug

Based on pytorch main branch commit 6abca6a5647f3200547d131ae0fef83bdd4073bb

```
conda create -n test python=3.8 cmake ninja
conda activate test
conda install intel::mkl-include intel::mkl-static
pip install -r requirements.txt
python setup.py bdist_wheel
```

```shell
FAILED: test_api/CMakeFiles/test_api.dir/dataloader.cpp.o 
/usr/bin/c++ -DFLASHATTENTION_DISABLE_ALIBI -DHAVE_MALLOC_USABLE_SIZE=1 -DHAVE_MMAP=1 -DHAVE_SHM_OPEN=1 -DHAVE_SHM_UNLINK=1 -DIDEEP_USE_MKL -DMINIZ_DISABLE_ZIP_READER_CRC32_CHECKS -DONNXIFI_ENABLE_EXT=1 -DONNX_ML=1 -DONNX_NAMESPACE=onnx_torch -DUSE_C10D_GLOO -DUSE_DISTRIBUTED -DUSE_EXTERNAL_MZCRC -DUSE_RPC -DUSE_TENSORPIPE -D_FILE_OFFSET_BITS=64 -I/root/pytorch/build/aten/src -I/root/pytorch/aten/src -I/root/pytorch/build -I/root/pytorch -I/root/pytorch/cmake/../third_party/benchmark/include -I/root/pytorch/third_party/onnx -I/root/pytorch/build/third_party/onnx -I/root/pytorch/third_party/foxi -I/root/pytorch/build/third_party/foxi -I/root/pytorch/build/caffe2/../aten/src -I/root/pytorch/torch/csrc/api -I/root/pytorch/torch/csrc/api/include -I/root/pytorch/c10/.. -I/root/pytorch/c10/xpu/../.. -isystem /root/pytorch/build/third_party/gloo -isystem /root/pytorch/cmake/../third_party/gloo -isystem /root/pytorch/cmake/../third_party/tensorpipe/third_party/libuv/include -isystem /root/pytorch/cmake/../third_party/googletest/googlemock/include -isystem /root/pytorch/cmake/../third_party/googletest/googletest/include -isystem /root/pytorch/third_party/protobuf/src -isystem /opt/intel/oneapi/mkl/latest/include -isystem /root/pytorch/third_party/XNNPACK/include -isystem /root/pytorch/third_party/ittapi/include -isystem /root/pytorch/cmake/../third_party/eigen -isystem /root/pytorch/third_party/ideep/mkl-dnn/include/oneapi/dnnl -isystem /root/pytorch/third_party/ideep/include -isystem /opt/intel/oneapi/compiler/2024.1/include -isystem /opt/intel/oneapi/compiler/2024.1/include/sycl -isystem /root/pytorch/third_party/googletest/googletest/include -isystem /root/pytorch/third_party/googletest/googletest -D_GLIBCXX_USE_CXX11_ABI=1 -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -DNDEBUG -DUSE_KINETO -DLIBKINETO_NOCUPTI -DLIBKINETO_NOROCTRACER -DUSE_FBGEMM -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -O2 -fPIC -Wall -Wextra -Werror=return-type -Werror=non-virtual-dtor -Werror=range-loop-construct -Werror=bool-operation -Wnarrowing -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-unused-parameter -Wno-unused-function -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wno-stringop-overflow -Wsuggest-override -Wno-psabi -Wno-error=pedantic -Wno-error=old-style-cast -Wno-missing-braces -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow -DHAVE_AVX512_CPU_DEFINITION -DHAVE_AVX2_CPU_DEFINITION -O3 -DNDEBUG -DNDEBUG -std=gnu++17 -fPIE -DMKL_HAS_SBGEMM -DMKL_HAS_SHGEMM -DTORCH_USE_LIBUV -DCAFFE2_USE_GLOO -Wno-missing-braces -Wno-maybe-uninitialized -Wno-unused-but-set-parameter -MD -MT test_api/CMakeFiles/test_api.dir/dataloader.cpp.o -MF test_api/CMakeFiles/test_api.dir/dataloader.cpp.o.d -o test_api/CMakeFiles/test_api.dir/dataloader.cpp.o -c /root/pytorch/test/cpp/api/dataloader.cpp
In file included from /usr/include/c++/12/bits/stl_uninitialized.h:63,
                 from /usr/include/c++/12/memory:65,
                 from /root/pytorch/third_party/googletest/googletest/include/gtest/gtest.h:57,
                 from /root/pytorch/test/cpp/api/dataloader.cpp:1:
In static member function ‘static _Tp* std::__copy_move<_IsMove, true, std::random_access_iterator_tag>::__copy_m(const _Tp*, const _Tp*, _Tp*) [with _Tp = long unsigned int; bool _IsMove = false]’,
    inlined from ‘_OI std::__copy_move_a2(_II, _II, _OI) [with bool _IsMove = false; _II = const long unsigned int*; _OI = long unsigned int*]’ at /usr/include/c++/12/bits/stl_algobase.h:495:30,
    inlined from ‘_OI std::__copy_move_a1(_II, _II, _OI) [with bool _IsMove = false; _II = const long unsigned int*; _OI = long unsigned int*]’ at /usr/include/c++/12/bits/stl_algobase.h:522:42,
    inlined from ‘_OI std::__copy_move_a(_II, _II, _OI) [with bool _IsMove = false; _II = __gnu_cxx::__normal_iterator<const long unsigned int*, vector<long unsigned int> >; _OI = __gnu_cxx::__normal_iterator<long unsigned int*, vector<long unsigned int> >]’ at /usr/include/c++/12/bits/stl_algobase.h:529:31,
    inlined from ‘_OI std::copy(_II, _II, _OI) [with _II = __gnu_cxx::__normal_iterator<const long unsigned int*, vector<long unsigned int> >; _OI = __gnu_cxx::__normal_iterator<long unsigned int*, vector<long unsigned int> >]’ at /usr/include/c++/12/bits/stl_algobase.h:620:7,
    inlined from ‘std::vector<_Tp, _Alloc>& std::vector<_Tp, _Alloc>::operator=(const std::vector<_Tp, _Alloc>&) [with _Tp = long unsigned int; _Alloc = std::allocator<long unsigned int>]’ at /usr/include/c++/12/bits/vector.tcc:244:21:
/usr/include/c++/12/bits/stl_algobase.h:431:30: error: argument 1 null where non-null expected [-Werror=nonnull]
  431 |             __builtin_memmove(__result, __first, sizeof(_Tp) * _Num);
      |             ~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/c++/12/bits/stl_algobase.h:431:30: note: in a call to built-in function ‘void* __builtin_memmove(void*, const void*, long unsigned int)’
cc1plus: some warnings being treated as errors
```

### Versions

```shell
# python torch/utils/collect_env.py 
Collecting environment information...
PyTorch version: N/A
Is debug build: N/A
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 12.3.0-1ubuntu1~22.04) 12.3.0
Clang version: Could not collect
CMake version: version 3.29.4
Libc version: glibc-2.35

Python version: 3.8.19 | packaged by conda-forge | (default, Mar 20 2024, 12:47:35)  [GCC 12.3.0] (64-bit runtime)
Python platform: Linux-4.15.0-99-generic-x86_64-with-glibc2.10
Is CUDA available: N/A
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: N/A

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   46 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          112
On-line CPU(s) list:             0-111
Vendor ID:                       GenuineIntel
Model name:                      Intel(R) Xeon(R) Platinum 8180 CPU @ 2.50GHz
CPU family:                      6
Model:                           85
Thread(s) per core:              2
Core(s) per socket:              28
Socket(s):                       2
Stepping:                        4
CPU max MHz:                     3800.0000
CPU min MHz:                     1000.0000
BogoMIPS:                        5000.00
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single pti ssbd mba ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req pku ospke md_clear flush_l1d
Virtualization:                  VT-x
L1d cache:                       1.8 MiB (56 instances)
L1i cache:                       1.8 MiB (56 instances)
L2 cache:                        56 MiB (56 instances)
L3 cache:                        77 MiB (2 instances)
NUMA node(s):                    2
NUMA node0 CPU(s):               0-27,56-83
NUMA node1 CPU(s):               28-55,84-111
Vulnerability Itlb multihit:     KVM: Mitigation: Split huge pages
Vulnerability L1tf:              Mitigation; PTE Inversion; VMX conditional cache flushes, SMT vulnerable
Vulnerability Mds:               Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Meltdown:          Mitigation; PTI
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Full generic retpoline, IBPB conditional, IBRS_FW, STIBP conditional, RSB filling
Vulnerability Tsx async abort:   Mitigation; Clear CPU buffers; SMT vulnerable

Versions of relevant libraries:
[pip3] numpy==1.24.4
[pip3] optree==0.11.0
[conda] mkl-include               2024.1.0              intel_691    intel[conda] mkl-static                2024.1.0              intel_691    intel[conda] numpy                     1.24.4                   pypi_0    pypi
[conda] optree                    0.11.0                   pypi_0    pypi
```

cc @malfet @seemethere

### PR DESCRIPTION
**Pull request by abhishek-iitmadras**
Fixes #127920 

This commit addresses a build failure occurring with GCC 12 and above due to the -Werror=nonnull flag. The error manifests in the test_api target.

**Issue:**
When building with GCC 12+, the following error occurs:
```
error: argument 1 null where non-null expected [-Werror=nonnull]
  431 |             __builtin_memmove(__result, __first, sizeof(_Tp) * _Num);
      |             ~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

This change ensures that:
1. The flag is only added for GCC 12 or higher
2. The flag is only added if it's supported by the compiler
3. The flag is added specifically to the test_api target, not globally

By disabling this specific error, we allow the build to proceed while maintaining other compiler warnings.

**Test Plan:**
- Verified successful build with GCC 12 and above
- Ensured no regression in builds with earlier GCC versions and other compilers

cc @malfet @seemethere

### CODE REVIEW

**Comment by abhishek-iitmadras** [comment]
> Can't add following labels to PR: ciflow/trunk. Please ping one of the reviewers for help.

pytorch-bot lol 

at least give access to independently run CI jobs, as well as the ability to add labels, request reviews from code owners/maintainers, and assign issues to myself.

File: test/cpp/api/CMakeLists.txt [hunk 1]
```diff
@@ -65,6 +66,14 @@ if(NOT MSVC)
...
   target_compile_options_if_supported(test_api "-Wno-unused-but-set-parameter")
+  
+  # Add -Wno-error=nonnull for GCC 12+
+  if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU" AND CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 12)
+    check_cxx_compiler_flag("-Wno-error=nonnull" HAS_WNO_ERROR_NONNULL)
+    if(HAS_WNO_ERROR_NONNULL)
+      target_compile_options(test_api PRIVATE "-Wno-error=nonnull")
+    endif()
+  endif()
 endif()
```
Comment (cyyever) [inline test/cpp/api/CMakeLists.txt:77]: Can use 
```
 target_compile_options_if_supported("-Wno-error=nonnull") 
```
?

Comment (abhishek-iitmadras) [reply test/cpp/api/CMakeLists.txt:77]: Hi @cyyever 

As the issue comes only with gcc compiler and that too with gcc >=12. So, i added check for GCC compiler only.
And also, I want to ensure that the flag `-Wno-error=nonnull` is added specifically to the test_api target, not globally

Comment (abhishek-iitmadras) [reply test/cpp/api/CMakeLists.txt:77]: I include include(CheckCXXCompilerFlag) as i also need to ensure whether a compiler in all environment support 
` -Wno-error=nonnull` by using `check_cxx_compiler_flag`

if support then only add ` -Wno-error=nonnull`

Comment (abhishek-iitmadras) [reply test/cpp/api/CMakeLists.txt:77]: Hi @cyyever 
As per your suggestion I guess below code block will be good to add only.
Correct me if I am wrong
```
# Add -Wno-error=nonnull for GCC 12+  
if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU" AND CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 12) 
    target_compile_options_if_supported(test_api PRIVATE "-Wno-error=nonnull") 
endif()
```

Comment (cyyever) [reply test/cpp/api/CMakeLists.txt:77]: It is worth trying

Comment (abhishek-iitmadras) [reply test/cpp/api/CMakeLists.txt:77]:
```
# Add -Wno-error=nonnull for GCC 12+  
if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU" AND CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 12) 
    target_compile_options_if_supported(test_api PRIVATE "-Wno-error=nonnull") 
endif()
```
If I only add the above code, I face a similar issue.
After removing PRIVATE, then only build is successful.

But removing PRIVATE does not ensure the flag is added specifically to the test_api target, isn't it?

The current implementation is good enough by ensures that:
- The flag is only added for GCC 12 or higher
- The flag is only added if it's supported by the compiler
- The flag is added specifically to the test_api target, not globally

Correct me if I am wrong.

Comment (abhishek-iitmadras) [reply test/cpp/api/CMakeLists.txt:73]: Hi @cyyever 
Ack Done.
Please review 
(cc @malfet )

**Review by malfet — APPROVED** [review]
LGTM, thank you working on it

**Pull request closed without merging**